File: P3Image.Domain/Services/SubCategoryService.cs
using System.Collections.Generic;
using System.Linq;
using System.Runtime.InteropServices;
using P3Image.Domain.Entities;
using P3Image.Domain.Interfaces.Repositories;
using P3Image.Domain.Interfaces.Services;

namespace P3Image.Domain.Services
{
    public class SubCategoryService : BaseService<SubCategory>, ISubCategoryService
    {
        public ISubCategoryRepository _rep;
        private IFieldRepository _fieldRep;

        public SubCategoryService(ISubCategoryRepository rep, IFieldRepository fieldRep)
            : base(rep)
        {
            _rep = rep;
            _fieldRep = fieldRep;
        }

        public void RemoveField(Field field)
        {
            _fieldRep.Remove(field);
        }

        public Field AddField(Field field)
        {
            return  _fieldRep.Add(field);
        }
        public Field EditField(Field field)
        {
            return _fieldRep.Update(field);
        }

        public List<Field> GetFields(SubCategory subCategory)
        {
           return  _fieldRep.GetAll().Where(x => x.SubCategoryId == subCategory.Id).ToList();
        }

        public SubCategory GetBySlug(string slugCategory, string slugSubCategory)
        {
            var subCategory = _rep.Search(x => x.Category.Slug == slugCategory && x.Slug == slugSubCategory).FirstOrDefault();

            if (subCategory != null) 
                subCategory.Fields = subCategory.Fields.OrderBy(x => x.Order).ToList();

            return subCategory;
        }
    }
}

File: P3Image.Web/Angular/Services/SubCategoryService.js

App.factory("subCategoryService", ["config", "httpService", "$q",
    function (config, httpService, $q) {
    	return {
    		save: function (entity) {
    			return httpService.post("api/subCategory/save", entity);
    		},
    		getAll: function () {
    			return httpService.get("api/subCategory/getAll");
    		},
    		getById: function (id) {
    			return httpService.get("api/subCategory/getById", { id: id });
    		},
    		getBySlug: function (slugCategory, slugSubCategory) {
    			return httpService.get("api/subCategory/getBySlug", {slugCategory: slugCategory, slugSubCategory: slugSubCategory
    		});
    				},
    		remove: function (entity) {
    			return httpService.post("api/subCategory/remove", entity);
    			},
    		addField: function (entity) {
    			return httpService.post("api/subCategory/addField", entity);
    			},
    		editField: function (entity) {
    			return httpService.post("api/subCategory/editField", entity);
    			},
    		removeField: function (entity) {
    			return httpService.post("api/subCategory/removeField", entity);
    			},
    		getFields: function (entity) {
    			return httpService.post("api/subCategory/getFields", entity);
    			},
    		validateSave: function (entity) {
    			var messages = [];
    			if (!entity.Category.Id) {
    				messages.push("Category must be selected");
    				}
    			if (!entity.Description || entity.Description.trim() === "") {
    				messages.push("Description must be filled");
    				}
    			if (!entity.Slug || entity.Slug.trim() === "") {
    				messages.push("Slug must be filled");
    				}
    			return messages;
    			}
    		}
    	}
    ]);

